Update readme to include guidelines for writing/updating errors (#73)
neo4j/docs-status-codes: README.adoc

* Avoid the phrase “It is recommended to” when proposing a solution.
* Capitalize all Cypher keywords, i.e., `MATCH` rather than `match` or `Match`.

> === Status code elements

Reviewer: Here we use the word 'elements', but in the text below and later we use 'components'

Each status code comprises the following components, some of which are documented only here in this repo.

[options="header",cols="h,2,2",]
|===
| Element
| Notification
| Error

| Status code (`Neo.[Class].[Subclass].[Name]`)
| `Neo.ClientNotification.Statement.CartesianProduct`.
| `Neo.DatabaseError.Statement.RemoteExecutionFailed`.

| Title/Description of the status code
| `This query builds a cartesian product between disconnected patterns.`
| `The database was unable to execute a remote part of the statement.`

| SeverityLevel
| One of `INFORMATION`\|`WARNING`
m| ERROR

| Category
| One of `DEPRECATION`\|`HINT`\|`UNSUPPORTED`\|`UNRECOGNIZED`\|`SECURITY`\|`GENERIC`\|`PERFORMANCE`\|`TOPOLOGY`
| N/A

| Description/Message
2+| Contains the main information, such as the problem, cause, consequences, a simple solution, or a reference to the Status Codes documentation.

| Example (only in Status Codes documentation)

> 2+| Contains one or more example queries to illustrate the possible scenarios when this notification code would be returned.

Reviewer: Currently, we do not have any examples of errors or error messages in here, do we plan on that and to what extent? I think we will start to just update error messages in the spreadsheet following the new guidelines and not add each error message as an example here? Maybe we need to do it eventually when we have all GQL codes.

Depending on the above, 'notification' should maybe be updated to whatever common word for notifications and errors we choose to use

| Suggestions for improvement (only in Status Codes documentation)
2+| Contains a possible solution for the provided example query.
|===

The following are guidelines on how to write each of the notification components.

==== StatusCode syntax

All Neo4j status codes follow the syntax `Neo.[Class].[Subclass].[Name]`.

[options="header",cols="h,1,2,3,3,2",]
|===
| StatusCode
| `Neo`
| `[Class]`
| `[Subclass]`
| `[Name]`
| Example

| Notification
| `Neo`
| `ClientNotification`
| One of `Statement`, `Procedure`, `Schema`, `Database`, `Security`, or `Request`
| Should be specific and explains what this code refers to, e.g., `ExhaustiveShortestPath`.
| `Neo.ClientNotification.Statement.ExhaustiveShortestPath`

| Error
| `Neo`
| One of `ClientError`, `TransientError`, `DatabaseError`
| One of `ChangeDataCapture`, `Cluster`, `Database`, `Fabric` (deprecated), `General`, `Procedure`, `Request`, `Routing`, `Schema`, `Security`, `Statement`, `Transaction`.
| Should be specific and explains what this code refers to, e.g., `RemoteExecutionFailed`.
| `Neo.DatabaseError.Statement.RemoteExecutionFailed`
|===

==== Notification title/Error description

> The notification title and the error description contain similar information.

Reviewer: I wonder if we should mention that the error description is normally not exposed to the user through the error (unless there is a fallback where it is used as error message) but only part of this documentation. What do you think?

Therefore, they follow the same guidelines.
Both should be brief (one short sentence), specific, and inform on the current situation and what code of the query triggered the notification/error (add it within backticks to show that it’s a code snippet).

> * Do not use exclamation marks.

Reviewer: I feel like this rule should maybe be under general rules as I think it should apply to the notification description/error message as well. Or do we think it is ok to use exclamation mark there?

* Use sentence case (you capitalize the first letter, proper nouns, and names, just like you would in a normal sentence) and punctuation.
* Avoid explaining the cause, consequences, or solution.

.Examples for notification title and error description
[options="header",cols="h,2,2",]
|===
| Example
| ⛔ Old title/description

> | ✅ New title/description

Reviewer: I know this was already here, but I wonder if we should write e.g. 'Bad' and 'Good' instead of 'Old' and 'New', given that we are not updating any of the existing notification titles, but only give guidelines for people that create new notifications?

Author: I didn't want to use `good` and `bad` deliberately. I've changed it to Don't/Do. What do you think?

Reviewer: Yes that works

| Notification title 1
| The provided pattern is unbounded, consider adding an upper limit to the number of node hops.
| The pattern `<pattern>` is unbounded.

| Notification title 2
| The provided label is not in the database.
| The label `Perso` does not exist.

| Notification title 3
| This feature is deprecated and will be removed in future versions.

> | (Deprecated) The repeated variable length relationship `r` is bound more than once.

Reviewer: I think we should remove this example as all deprecations have the above title and that cannot be changed (it is breaking to change in 5.x and the title will anyway be removed in 6.x). The others cannot be changed either, but can more be seen as general examples

|===

==== Notification description/error message

The Notification description and error message should contain the most important information for the user.
They should be brief, scannable, specific, and contain the following details (if applicable):

* Problem -- what happened (if not already mentioned in the title).
* Cause -- what triggered the code to be returned.
* Consequences (for notifications only) -- why it might be a problem.

> * A simple solution or a reference to the docs where further examples and details are provided, using the sentence: _See Status Codes documentation for suggestions._

Reviewer: ```suggestion
* A simple solution.
```

We can remove this now when you are anyway updating the guidelines.

[options="header",cols="h,2,2,2"]
|===
| Element
| More information
| Notification
| Error

| StatusCode (`Neo.[Class].[Subclass].[Name]`)
| See <<statuscode-syntax, StatusCode syntax>>.
| For example, `Neo.ClientNotification.Statement.CartesianProduct`.
| For example, `Neo.DatabaseError.Statement.RemoteExecutionFailed`.

| Notification title/ Error description (Will be removed in 6.0)
| See <<notification-title-error-description, Notification title/Error description>>.
| For example, `This query builds a cartesian product between disconnected patterns.`
| For example, `The database was unable to execute a remote part of the statement.`

| SeverityLevel
|
| One of `INFORMATION`\|`WARNING`
m| ERROR

| Category
|
| One of `DEPRECATION`\|`HINT`\|`UNSUPPORTED`\|`UNRECOGNIZED`\|`SECURITY`\|`GENERIC`\|`PERFORMANCE`\|`TOPOLOGY`
| N/A

| Notification description (Will be removed in 6.0)
| See <<notification-description-error-message, Notification description>>.
| Contains the main information, such as the problem, cause, consequences, a simple solution, or a reference to the Status Codes documentation.
| N/A

> | Message (A new component that will replace the notification title and description, and the error description in 6.0)

Reviewer: Well, this is almost correct. Errors do have messages already today but we plan to update those in 6.0

| `[Class]`
| `[Subclass]`
| `[Name]`
| Example

| Notification
| `Neo`
| `ClientNotification`
| One of `Statement`, `Procedure`, `Schema`, `Database`, `Security`, or `Request`
| Should be specific and explains what this code refers to, e.g., `ExhaustiveShortestPath`.
| `Neo.ClientNotification.Statement.ExhaustiveShortestPath`

| Error
| `Neo`
| One of `ClientError`, `TransientError`, `DatabaseError`
| One of `ChangeDataCapture`, `Cluster`, `Database`, `Fabric` (deprecated), `General`, `Procedure`, `Request`, `Routing`, `Schema`, `Security`, `Statement`, `Transaction`.
| Should be specific and explains what this code refers to, e.g., `RemoteExecutionFailed`.
| `Neo.DatabaseError.Statement.RemoteExecutionFailed`
|===

[[notification-title-error-description]]
==== Notification title/error description

The notification title and the error description contain similar information.
Therefore, they follow the same guidelines.
Both should be brief (one short sentence), specific, and inform on the current situation and what code of the query triggered the notification/error (add it within backticks to show that it’s a code snippet).
Avoid explaining the cause, consequences, or solution.

> The notification title and description will be merged in 6.0, and the error description will be replaced by the message.

Reviewer 2: ```suggestion
The notification title and description will be merged in 6.0. The error description is in 5.x just used as a fallback-error message in some specific cases and in 6.0 they will no longer be used.
```

See <<message, Message>>.

.Examples for notification title and error description
[options="header",cols="h,2,2",]
|===
| Example
| ⛔ Don't
| ✅ Do

| Notification title 1
| The provided pattern is unbounded, consider adding an upper limit to the number of node hops.
| The pattern `<pattern>` is unbounded.

| Notification title 2
| The provided label is not in the database.
| The label `Perso` does not exist.
|===

[[notification-description]]
==== Notification description

The Notification description should contain the most important information for the user.
They should be brief, scannable, specific, and contain the following details (if applicable):

* Cause -- what triggered the code to be returned.
* Consequences (for notifications only) -- why it might be a problem.
* A simple solution if possible.

> The notification description will be removed in 6.0, and the message will be used instead.

Reviewer 2: With the change above, this line can be removed.

Avoid explaining the cause, consequences, or solution.
The notification title and description will be replaced by a single field `message` in 6.0.
The error description in 5.x is used just as a fallback error message in some specific cases, and in 6.0, it will no longer be used.
See <<message, Message>>.

.Examples for notification title and error description
[options="header",cols="h,2,2",]
|===
| Example
| ⛔ Don't
| ✅ Do

| Notification title 1
| The provided pattern is unbounded, consider adding an upper limit to the number of node hops.
| The pattern `<pattern>` is unbounded.

| Notification title 2
| The provided label is not in the database.
| The label `Perso` does not exist.
|===

[[notification-description]]
==== Notification description

The Notification description should contain the most important information for the user.
They should be brief, scannable, specific, and contain the following details (if applicable):

* Cause -- what triggered the code to be returned.

> * Consequences (for notifications only) -- why it might be a problem.

Reviewer: ```suggestion
* Consequences -- why it might be a problem.
```

Not needed as this is a notification-only section now

| A variable length relationship that is bound more than once does not return any result.

| Notification description 3
| One of the labels in your query is not available in the database, make sure you didn’t misspell it or that the label is available when you run this statement in your application (the missing label name is: Perso)
| Non-existing labels yield no result. Verify that the label is spelled correctly.
|===

[[message]]
==== Message

In 6.0, the notifications and errors will have only a message, which will be used instead of the notification title and description, and the error description and message.

The message should follow the same guidelines as the notification description, namely:

The message should contain the most important information for the user.
It should be brief, scannable, specific, and contain the following details (if applicable):

* Problem -- what happened and what code of the query triggered the notification/error (add it within backticks to show that it’s a code snippet).
* Cause -- what triggered the code to be returned.

> * Consequences (for notifications only) -- why it might be a problem.

Reviewer: ```suggestion
```

Suggestions for improvement::

Give a possible solution for the provided example query.

> ====

Reviewer: Github would not let me mark that line, but there is a typo in Neo4j on line 226 🙈

For example:

[options="header",cols="h,2,2",]
|===
| Example
| ⛔ Old example

> | ✅ New example

Reviewer: Same as above, I think 'Old' and 'New' example does not make much sense here, as we only have new examples.

If for example, you need the children and the children's parents, you can improve this query by rewriting it to the following:
+
[source, cypher, role="noplay"]
----
MATCH (c:Child)-[:ChildOf]->(p:Parent) RETURN c, p
----
====
a| .Cartesian product
====
Query::
+
[source, cypher, role="noplay"]
----
MATCH (c:Child), (p:Parent) RETURN c, p
----

Description of the returned code::
The disconnected patterns `$param` build a cartesian product. A cartesian product may produce a large amount of data and slow down query processing. See Status Codes documentation for suggestions.

Suggestions for improvement::
If you only need the children and the children's parents, and not all combinations between them, add `[:ChildOf]` between the `Child` and the `Parent` nodes:

[source, cypher, role="noplay"]
----
MATCH (c:Child)-[:ChildOf]->(p:Parent) RETURN c, p
----
====

| Error example
a| .Using a list range operator `[ ]` without a start or an end value.
====
Query::
+
[source, cypher, role="noplay"]
----
RETURN range(0,10)[..]
----

Description of the returned code::

> The statement is syntactically valid, but expresses something that the database cannot do.

Reviewer: If we are planning on adding error examples to this docs manual, I think we should omit this for them as the description is connected to the Neo4j status code and thus already documented above the example

> Message of the returned code::

Reviewer: This is not mentioned above where we explain how an example should look (after "They should look similar to the following:")
